### ISSUE TITLE
HTTP Basic Auth without Forced Challenge

### ISSUE THREAD

**Issue by ravage84** [cakephp/authentication#497]
CakePHP 3.x
Authentication 1.x

We currently develop a RESTful API with HTTP basic auth with several endpoints.
Some endpoints need authentication, some do not.

I'm aware of the following issues and that we could use the URL prefix to handle whether to auth or not.

https://github.com/cakephp/authentication/issues/258
https://github.com/cakephp/authentication/issues/381

But I found a different solution, which is rather simple and provides a more flexible or let's say more selective solution.

```` php
    public function getAuthenticationService(ServerRequestInterface $request, ResponseInterface $response)
    {
        $service = new AuthenticationService();
        $service->loadAuthenticator(
            'OurApi.HttpBasic',
            [
                'className' => ApiHttpBasicAuthenticator::class,
            ]
        );

        // snip...

        return $service;
    }
````

So instead of using ``Authentication\Authenticator\HttpBasicAuthenticator``one could implement a simpler authenticator class using the same code as the original class, except the following:

1. Not extending the ``StatelessInterface`` 
2. ``unauthorizedChallenge()`` method
3. ``loginHeaders()`` method

This prevents any challenge being sent if the auth credentials aren't provided.

Then, one can load the ``AuthenticationComponent`` like so:

```` php
    public function initialize()
    {
        parent::initialize();

        $this->loadComponent(
            'Authentication.Authentication',
            [
                'requireIdentity' => true,
            ]
        );
    }
````

Lastly, if some endpoints do not need any auth, they can be excluded by using:

```` php
    public function beforeFilter(Event $event)
    {
        $this->Authentication->allowUnauthenticated(['ping']);

        return parent::beforeFilter($event);
    }
````

1. Is there a particular reason why this is not documented? E.g. not recommended.
2. Or at least provided by the plugin?

If not, would we want to have such an alternative HTTP basic auth authenticator?

**Comment by ravage84** [comment]
Another use case for a non-challenging HTTP Basic Authenticator class is when you want to use a custom error handler, instead of letting the ``AuthenticationMiddleware`` [handle it](https://github.com/cakephp/authentication/blob/1.4.2/src/Middleware/AuthenticationMiddleware.php#L97-L105).

**Comment by markstory** [comment]
If you are using basic auth and don't force a challenge how does the client know to prompt the user for basic authentication credentials?

**Comment by ravage84** [comment]
@markstory first of all, I simply don't want he middleware to do the challenge as code after it should be reached.
Some code after that could do it.

Second even if there was no challenge, a client could still know it by documentation, e.g. an OpenAPI spec.

Lastly, an endpoint should send a 401 status if there was no kind of valid authentication (there could be multiple), nonetheless.

**Comment by ADmad** [comment]
I see that currently when using the `HttpBasicAuthenticator` a response with `WWW-Authenticate` header is thrown for every failure case. As per the HTTP Auth specs the challenge headers should be sent to the client only when no credentials are provided, not for the case when the credentials are invalid.

So probably this is the change that's required:

```diff
diff --git a/src/AuthenticationService.php b/src/AuthenticationService.php
index b25c126..04b24f0 100644
--- a/src/AuthenticationService.php
+++ b/src/AuthenticationService.php
@@ -19,6 +19,7 @@ namespace Authentication;
 use Authentication\Authenticator\AuthenticatorCollection;
 use Authentication\Authenticator\AuthenticatorInterface;
 use Authentication\Authenticator\PersistenceInterface;
+use Authentication\Authenticator\Result;
 use Authentication\Authenticator\ResultInterface;
 use Authentication\Authenticator\StatelessInterface;
 use Authentication\Identifier\IdentifierCollection;
@@ -187,7 +188,10 @@ class AuthenticationService implements AuthenticationServiceInterface
                 return $this->_result = $result;
             }
 
-            if ($authenticator instanceof StatelessInterface) {
+            if (
+                $authenticator instanceof StatelessInterface
+                && $result->getStatus() === Result::FAILURE_CREDENTIALS_MISSING
+            ) {
                 $authenticator->unauthorizedChallenge($request);
             }
         }
```

**Comment by ADmad** [comment]
Okay my suggested patch above doesn't work since `HttpBasicAuthenticator` returns a result with status `FAILURE_CREDENTIALS_MISSING` even when `PHP_AUTH_USER` and `PHP_AUTH_PW` env vars are present. So that's another this that needs to be addressed.

**Comment by ravage84** [comment]
@ADmad the wording for whether a challenge will be sent or not seems to be "can", not even "should":

> Upon receipt of a request for a URI within the protection space that
   lacks credentials, the server **can** reply with a challenge using the
   401 (Unauthorized) status code ([RFC7235], Section 3.1) and the
   WWW-Authenticate header field ([RFC7235], Section 4.1).

https://datatracker.ietf.org/doc/html/rfc7617

Anyway, I didn't want to propose changing the current ``HttpBasicAuthenticator``, but instead adding another one that does not enforce the challenge.
Some other part of an application's code can still enforce that (which would be sensible in most cases).

**Comment by ADmad** [comment]
Since sending a challenge for invalid credentials at least is simply wrong, I have opened a PR to address that. 

Based on the wording in the RFC, since sending the challenge even when credentials are missing seems optional, we can just add a config to `HttpBasicAuthenticate` to skip throwing the challenge.

**Comment by ADmad** [comment]
Closed by #498
